src/core/lib/event_engine/ares_resolver.cc

```diff
@@ -809,6 +854,51 @@ void AresResolver::OnTXTDoneLocked(void* arg, int status, int /*timeouts*/,
       });
 }
 
+#ifdef GRPC_ENABLE_FORK_SUPPORT
+
+std::weak_ptr<AresResolver::ReinitHandle> AresResolver::GetReinitHandle() {
+  grpc_core::MutexLock lock(&reinit_handle_mu_);
+  if (reinit_handle_ == nullptr) {
+    reinit_handle_ = std::make_shared<ReinitHandle>(this);
+  }
+  return reinit_handle_;
+}
+
+void AresResolver::Reinitialize() {
+  auto self = RefIfNonZero();
+  if (self == nullptr) {
+    return;
+  }
+  // Let the callbacks know
+  // TODO (eostroukhov): Should it be merged with Orphan? Current differences:
```

Comment:
It seems like you could refactor this into a common `ShutdownLocked()` method, which could be called both from `Orphan()` and here.  You can pass in the log message, and the code in `Orphan()` can set `shutting_down_` true before it calls `ShutdownLocked()`.